architecture-decisions/0003-preservation.md

```diff
@@ -0,0 +1,67 @@
+   1. We will package the bag directory in a Tar archive, compressed with GZip.
+   1. A link to the Tar/GZipped bag file will be added to the object's `file_metadata` property.
+   1. We will calculate the MD5 checksum of the Tar/GZipped bag and include it in the `file_metadata`.
+   1. We will assign the `file_metadata` a PCDM Use value of `pcdm:preservationArchive`.
+1. Transfering bags
+   1. After we transfer a bag to cloud storge, we will verify that its MD5 checksum matches the locally-
+      calculated value.
+   1. We will keep only the most recent bag for an object, overwriting the bag with a new version when the
+      object is updated.
+1. Deleting bags
+   1. There will not be an automated way to delete a bag.  Any deletions will need to be done manually.
+   1. When an object is deleted, the preservation bag will not be deleted.
+
+## Consequences
+
+1. Creating bags
+   1. There will be a delay between when objects are created and/or updated and when they are bagged.
+   1. The scanning frequency will determine how often bags may be redundantly created and uploaded.
+1. Packaging bags
+   1. Exporting all of the metadata and linked objects with a bag will make it a better preservation package
+      because it will include many of the objects that provide context for the object.  However, these will
+      also make the bag larger.
+   1. Packaging the bags in a Tar/GZip archive will make it easier to work with, transfer, and verify. But
+      it will also make it a larger single file, increasing the risk of problems during transfer or copying.
+      Compression will reduce file size slightly, but magnify the impact of transfer or copying errors.
+   1. Linking from the object to the bag will make it easy to find the bag that preserves an object.  But it
+      will also remove the place where the bag is recorded when an object is deleted.
+1. Transfering bags
+   1. Storing bags in cloud storage makes it much less likely that they will be accidentally corrupted or
+      deleted by reducing the number of staff who have access to the bag storage.
+   1. Storing bags in cloud storage will store them in a geographically separate area using a separate
+      technology stack, reducing vulnerability to natural disaster and malicious attack.
+   1. Using MD5 checksums to verify uploading will reduce the possibility of corruption or truncation in 
+      transfer.  There is a very small chance that the archive might be compromised with a malicious file
+      with a colliding MD5 checksum.  But the risk of this kind of attack is very low.
+1. Deleting bags
+   1. Deleting bags manually means the last preserved state of the object will persist after it is deleted.
+   1. Not deleting bags automatically means that we may accrue bags for deleted objects and waste space. But
+      it also means that preservation bags can be used to restore accidentally deleted objects.
```

Comment:
Does this imply that bags are a separate mechanism for disaster recovery?  Are these given preferential treatment to restoring the entire system state using PostgreSQL backups?

If so, I would also be concerned about this serving the function of proper versioning support.

Reply:
I think that our backups are good for certain scenarios, such as corruption/failure of the database or storage, and (assuming it's noticed quickly) recovering from accidental deletion.

The bags are intended for longer-term recovery and larger-scale disaster recovery (up to and including complete destruction of the datacenter, failure of the entire software stack, etc.).

Versioning is a tricky issue, and I haven't really heard our stakeholders ask for it.  But it's a tool that's available and might be appropriate for addressing certain needs.